drivers: serial: Add rpmsg driver

Ayush1325 opened the pull request:
A lot of BeagleBoard.org boards contain SOCs with co-processors such as
M4F in AM62x (PocketBeagle 2 [0]), R5s in AM67A (BeagleY-AI [1]). In such
targets, the application is normally loaded by the Linux host using
remoteproc.

There have been some out of tree patches to have micropython and other
Zephyr applications output to allow having a console without requiring
manual connections. This patch attempts to provide a more concrete and
upstream way to have that functionality.

Most of the code is copied from the openamp_rsc_table sample.

Current rpmsg related APIs use system heap, which is not the best. I can
add additional APIs to allow creating rpmsg console using pre-allocated
structures.

Only implemented poll_out

Tested on PocketBeagle 2 M4F core.

[0]: https://docs.zephyrproject.org/latest/boards/beagle/pocketbeagle_2/doc/index.html
[1]: https://docs.zephyrproject.org/latest/boards/beagle/beagley_ai/doc/index.html

Succeeds https://github.com/zephyrproject-rtos/zephyr/pull/96381

cc @arnopo

Ayush1325 committed:
drivers: serial: Add RPMSG UART

A lot of BeagleBoard.org boards contain SOCs with co-processors such as
M4F in AM62x (PocketBeagle 2 [0]), R5s in AM67A (BeagleY-AI [1]). In such
targets, the application is normally loaded by the Linux host using
remoteproc.

There have been some out of tree patches to have micropython and other
Zephyr applications output to allow having a console without requiring
manual connections. This patch attempts to provide a more concrete and
upstream way to have that functionality.

The implementation uses existing rpmsg service subsystem.

Only implemented poll_out

Tested on PocketBeagle 2 M4F core.

[0]: https://docs.zephyrproject.org/latest/boards/beagle/pocketbeagle_2/doc/index.html
[1]: https://docs.zephyrproject.org/latest/boards/beagle/beagley_ai/doc/index.html

Signed-off-by: Ayush Singh <ayush@beagleboard.org>

nordicjm commented:
I actually made my own for this too, then I noticed that there already is an in-tree IPC for logging - see https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/subsys/logging/multidomain for a sample

glneo commented:
The shell backend is too high level and most applications do not use it. But the UART/serial layer is too low level, usually working uncooked and with single chars. RPMSG supports writing in longer bursts and the overhead of one message per char would be huge.

The TTY/Console level would be ideal. Maybe try to see what a replacement for `tty_{read,write}()` using RPMSG would look like. Would also match the other side of this as the Linux kernel side is also a tty_driver `rpmsg_tty.c`, not a UART driver.

Ayush1325 commented:
> The shell backend is too high level and most applications do not use it. But the UART/serial layer is too low level, usually working uncooked and with single chars. RPMSG supports writing in longer bursts and the overhead of one message per char would be huge.

For the interrupt and async usecases, we can use ringbuf + work queue, but I thought for basic poll-out, it should probably be direct writes.
 
> The TTY/Console level would be ideal. Maybe try to see what a replacement for `tty_{read,write}()` using RPMSG would look like. Would also match the other side of this as the Linux kernel side is also a tty_driver `rpmsg_tty.c`, not a UART driver.

Well, the zephyr console subsystem still only works with single chars. I did create a rpmsg console driver in a previous PR: https://github.com/zephyrproject-rtos/zephyr/pull/96381

Ayush1325 committed:
ipc: rpmsg_service: Add openamp resource table support

Use resource table when CONFIG_OPENAMP_RSC_TABLE is enabled.

Tested with PocketBeagle 2 (M4 core running Zephyr)

Signed-off-by: Ayush Singh <ayush@beagleboard.org>

Ayush1325 commented:
> > The current backend for zephyr remote and host does not seem to work with a Linux host. So add a new backend.
> 
> This isn't the right way to go about this. Have you looked into why the current remote backend doesn't work with Linux?

Well, I think the problem is that the rpmsg backend is not really using the resource table, since it seems to be designed and tested with zephyr master only. But I can be wrong.

Ayush1325 committed:
boards: beagleboard: pocketbeagle_2: m4: Add rpmsg uart

- Add rpmsg uart in dt.
- Not enabling it by default yet.

Signed-off-by: Ayush Singh <ayush@beagleboard.org>

Ayush1325 commented:
> That seems like a reasonable guess at what the problem might be. How about we add the resource table handling code to the current rpmsg backend conditionally on `CONFIG_OPENAMP_RSC_TABLE` or `OPENAMP_RSC_TABLE_NUM_RPMSG_BUFF > 0`?

I have added a kconfig option for enabling linux specific stuff in the existing backend now. It is currently disabled by default.

jadonk reviewed (dismissed):
This seems reasonable and small to enable a lot of development capability. I could echo some of the other feedback, but I don't see a fundamental issue to hold this up.

arnopo reviewed (commented):
My apologies for this late review; it slipped off my radar.

The implementation looks good to me. I have just one main comment regarding the CONFIG, expecting that it does has been already discuted 

Since I’m late, it’s fair if my comments are ignored.

Ayush1325 commented on subsys/ipc/rpmsg_service/Kconfig:
No longer required

glneo reviewed (changes requested):
Fix CI failure

arnopo reviewed (approved):
LGTM
